Cache StringNames and NodePaths in implicit casts from string in C#

Joy-less opened the pull request:
This pull request addresses https://github.com/godotengine/godot-proposals/discussions/10826 (please take a look).
To summarise, StringNames and NodePaths are now cached when implicitly casting from strings to avoid allocations like:
```cs
Input.IsActionPressed("action"); // Created a StringName.
GetNode("path/to/node"); // Created a NodePath.
```

This pull request adds a [FusionCache](https://github.com/ZiggyCreatures/FusionCache) dependency since there is no cache built-in to C#. This could be replaced with a [MemoryCache](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.caching.memory.memorycache) dependency, or an in-house solution could be created using a `ConcurrentDictionary`.

In my opinion, StringNames (and NodePaths) do not make sense in C# and should be removed from the API. The reason is the garbage collection spike issues I detailed in the [discussion](https://github.com/godotengine/godot-proposals/discussions/10826). There are very few benefits to StringNames since identifiers are generally less than 20 characters long, and NodePaths don't seem to provide any value.

That being said, as long as StringNames and NodePaths remain in the API, implicit allocations should not exist, especially when it's so common to use string literals as StringNames and NodePaths.

If this pull request is added, the developer can still opt for a non-cached StringName or NodePath using `new StringName(string)` or `new NodePath(string)`.

Another related issue is there is currently no equality operator between StringName and string, so comparing them will result in the string being converted to a StringName, which is not ideal. It may be a good idea to add an `==` operator to StringName which accepts a string and compares them by converting the StringName to a string (instead of converting the string to a StringName).

By no means is this the only solution to the problem - I am open to discussions about better alternatives.

Joy-less commented:
I ran some benchmarks using [BenchmarkDotNet.Godot](https://github.com/ZerxZ/BenchmarkDotNet.Godot) with alternative caching mechanisms:

| Method                         | Mean      | Error    | StdDev   | Gen0       | Gen1      | Allocated  |
|------------------------------- |----------:|---------:|---------:|-----------:|----------:|-----------:|
| NoneTest                       | 142.99 ms | 2.141 ms | 2.003 ms |  3000.0000 | 2750.0000 | 19200760 B |
| FusionCacheTest                |  33.66 ms | 0.072 ms | 0.064 ms | 12200.0000 |         - | 38400494 B |
| MemoryCacheTest                |  19.34 ms | 0.045 ms | 0.040 ms |  4062.5000 |         - | 12800410 B |
| MemoryCacheStaticTest          |  17.63 ms | 0.074 ms | 0.058 ms |          - |         - |      410 B |
| ConcurrentDictionaryTest       |  16.55 ms | 0.022 ms | 0.018 ms |  4062.5000 |         - | 12800410 B |
| ConcurrentDictionaryStaticTest |  16.57 ms | 0.010 ms | 0.008 ms |          - |         - |      410 B |
| DictionaryWithLockTest         |  16.95 ms | 0.251 ms | 0.235 ms |          - |         - |      410 B |

<details>
<summary>Benchmark Code</summary>

```cs
#nullable enable

using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Threading;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Godot.Attributes;
using BenchmarkDotNet.Godot.Attributes.Jobs;
using Godot;
using Microsoft.Extensions.Caching.Memory;
using ZiggyCreatures.Caching.Fusion;

[MemoryDiagnoser, GodotSimpleJob]
public partial class StringNameBenchmark : Node {
    [Export] public int TestInteger = 53;

    public const int Count = 200_000;

    private readonly FusionCache FusionCache = new(new FusionCacheOptions());
    private readonly MemoryCache MemoryCache = new(new MemoryCacheOptions());
    private readonly ConcurrentDictionary<string, StringName> ConcurrentDictionary = [];
    private readonly Dictionary<string, StringName> Dictionary = [];
    private readonly Lock Lock = new();
    private string TargetString = "TestInteger";

    [GodotBenchmark]
    public void NoneTest() {
        for (int Counter = 0; Counter < Count; Counter++) {
            Get(TargetString);
        }
    }
    [GodotBenchmark]
    public void FusionCacheTest() {
        for (int Counter = 0; Counter < Count; Counter++) {
            Get(FusionCache.GetOrSet(TargetString, _ => (StringName)TargetString));
        }
    }
    [GodotBenchmark]
    public void MemoryCacheTest() {
        for (int Counter = 0; Counter < Count; Counter++) {
            Get(MemoryCache.GetOrCreate(TargetString, _ => (StringName)TargetString));
        }
    }
    [GodotBenchmark]
    public void MemoryCacheStaticTest() {
        for (int Counter = 0; Counter < Count; Counter++) {
            Get(MemoryCache.GetOrCreate(TargetString, static (ICacheEntry Entry) => (StringName)(string)Entry.Key));
        }
    }
    [GodotBenchmark]
    public void ConcurrentDictionaryTest() {
        for (int Counter = 0; Counter < Count; Counter++) {
            Get(ConcurrentDictionary.GetOrAdd(TargetString, _ => (StringName)TargetString));
        }
    }
    [GodotBenchmark]
    public void ConcurrentDictionaryStaticTest() {
        for (int Counter = 0; Counter < Count; Counter++) {
            Get(ConcurrentDictionary.GetOrAdd(TargetString, static (string Key) => (StringName)Key));
        }
    }
    [GodotBenchmark]
    public void DictionaryWithLockTest() {
        for (int Counter = 0; Counter < Count; Counter++) {
            lock (Lock) {
                if (!Dictionary.TryGetValue(TargetString, out StringName? Result)) {
                    Dictionary[TargetString] = (StringName)TargetString;
                }
                Get(Result);
            }
        }
    }
}
```
</details>

`None` is the worst, with large Generation 1 heap allocations and the slowest performance.

`FusionCache` is the second worst, likely because of two things:
- It appears to be designed for ASP.NET to cache very slow things like web requests, rather than class allocations.
- It has to allocate a delegate for the factory function.

I think the best option is to change to `MemoryCache` and use it like so:
```cs
string TargetString = "TestNumber";
StringName TargetStringName = StringNameCache.GetOrCreate(TargetString, static (ICacheEntry Entry) => new StringName((string)Entry.Key));
```
The static lambda avoids allocating any delegates. The unboxing should not be a problem since `string` is a reference type.

The implicit cast would end up looking like this:
```cs
public static implicit operator StringName(string from) => _stringNameCache.GetOrCreate(from, static (ICacheEntry entry) => new StringName((string)entry.Key))!;
```

jodydonetti commented:
Hi all, FusionCache creator here: let me know if you need anything.

Just FYI: `MemoryCache` does **not** have [cache stampede](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/CacheStampede.md) protection.

Also, after months of work, FusionCache V2 is about to be released in the next few weeks with a lot of extra features.

Here is the [last preview](https://github.com/ZiggyCreatures/FusionCache/releases/tag/v2.0.0-preview-4).

Hope this helps!

jodydonetti commented:
I don't know that much about Godot, so take this with a pinch of salt, but: for something so core and so core and so low level I would probably not use FusionCache, nor MemoryCache (its API surface area is not really great and allocates a disposable for every `Set` operation, for example) nor anything like that at all.

I think I'd create something uber low level and 100% specialized for this exact purpose, basically something like a `Dictionary`/`ConcurrentDictionary` + massive use of `ReadOnlySpan<T>` or similar + some form of low level locking, if needed at all: for example if an extra string allocation in the worst case scenario is not the end of the world I may not even put a locking mechanism and just go with the native dictionary `GetOrAdd` method.

To be clear: I'm saying this not because FusionCache is bad, I mean, I created it (and FWIW I think it's really good 😬).

The thing is that it does **a lot** like [L1+L2](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/CacheLevels.md), [backplane](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/Backplane.md), [fail-safe](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/FailSafe.md), [eager-refresh](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/EagerRefresh.md), [conditional refresh](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/ConditionalRefresh.md), [auto-recovery](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/AutoRecovery.md) and way more, and if what is needed here is "just" an in-memory dictionary with some concurrent access mechanism, that may be overkill.

Hope this helps.

jodydonetti commented:
Hi @Joy-less 

> * It has to allocate a delegate for the factory function.

I'm not sure I got this 100% right, but to be clear: FusionCache can of course work with a static delegate and no allocation at all.
To be more explicit, I have created some benchmarks to test exactly this with a so called "read happy path", and the memory allocated is exactly zero bytes.

Hope this helps.

Joy-less commented:
I have changed the cache again to [`BitFaster.Caching :: ConcurrentLru`](https://github.com/bitfaster/BitFaster.Caching) after discussing with @jodydonetti. ([RCache](https://github.com/dotnet/extensions/issues/4766)), a new .NET caching library, has been in development and is [based on BitFaster](https://github.com/dotnet/extensions/issues/4766#:~:text=RCache%20implementation%20is%20based%20on%20open%20source%20library%20BitFaster.).
`ConcurrentLru` is essentially a `ConcurrentDictionary` with a hard capacity, a least-recently-used eviction policy, and an optional [expire after access](https://github.com/bitfaster/BitFaster.Caching/wiki/ConcurrentLru:-Time%E2%80%90based-eviction) time-eviction policy. This makes it very fast for this use case.

<details>
<summary>Benchmark code</summary>

```cs
#nullable enable

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Threading;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Godot.Attributes;
using BenchmarkDotNet.Godot.Attributes.Jobs;
using BitFaster.Caching;
using BitFaster.Caching.Lfu;
using BitFaster.Caching.Lru;
using Godot;
using Microsoft.Extensions.Caching.Memory;
using ZiggyCreatures.Caching.Fusion;

[MemoryDiagnoser, GodotSimpleJob]
public partial class StringNameBenchmark : Node {
    [Export] public int TestInteger = 53;

    public const int Count = 200_000;

    private readonly FusionCache FusionCache = new(new FusionCacheOptions());
    private readonly MemoryCache MemoryCache = new(new MemoryCacheOptions());
    private readonly ConcurrentLru<string, StringName> ConcurrentLru = new(capacity: 1_000);
    private readonly ConcurrentLfu<string, StringName> ConcurrentLfu = new(capacity: 1_000);
    private readonly ICache<string, StringName> ExpiringConcurrentLru = new ConcurrentLruBuilder<string, StringName>()
        .WithCapacity(1_000)
        .WithExpireAfterAccess(TimeSpan.FromSeconds(30))
        .Build();
    private readonly ConcurrentDictionary<string, StringName> ConcurrentDictionary = [];
    private readonly Dictionary<string, StringName> Dictionary = [];
    private readonly Lock Lock = new();
    private string TargetString = "TestInteger";

    [GodotBenchmark]
    public void NoneTest() {
        for (int Counter = 0; Counter < Count; Counter++) {
            Get(TargetString);
        }
    }
    [GodotBenchmark]
    public void FusionCacheTest() {
        for (int Counter = 0; Counter < Count; Counter++) {
            Get(FusionCache.GetOrSet(TargetString, _ => (StringName)TargetString));
        }
    }
    [GodotBenchmark]
    public void MemoryCacheTest() {
        for (int Counter = 0; Counter < Count; Counter++) {
            Get(MemoryCache.GetOrCreate(TargetString, _ => (StringName)TargetString));
        }
    }
    [GodotBenchmark]
    public void MemoryCacheStaticTest() {
        for (int Counter = 0; Counter < Count; Counter++) {
            Get(MemoryCache.GetOrCreate(TargetString, static (ICacheEntry Entry) => (StringName)(string)Entry.Key));
        }
    }
    [GodotBenchmark]
    public void ConcurrentDictionaryTest() {
        for (int Counter = 0; Counter < Count; Counter++) {
            Get(ConcurrentDictionary.GetOrAdd(TargetString, _ => (StringName)TargetString));
        }
    }
    [GodotBenchmark]
    public void ConcurrentDictionaryStaticTest() {
        for (int Counter = 0; Counter < Count; Counter++) {
            Get(ConcurrentDictionary.GetOrAdd(TargetString, static (string Key) => (StringName)Key));
        }
    }
    [GodotBenchmark]
    public void DictionaryWithLockTest() {
        for (int Counter = 0; Counter < Count; Counter++) {
            lock (Lock) {
                if (!Dictionary.TryGetValue(TargetString, out StringName? Result)) {
                    Dictionary[TargetString] = (StringName)TargetString;
                }
                Get(Result);
            }
        }
    }
    [GodotBenchmark]
    public void ConcurrentLruTest() {
        for (int Counter = 0; Counter < Count; Counter++) {
            Get(ConcurrentLru.GetOrAdd(TargetString, static (string TargetString) => (StringName)TargetString));
        }
    }
    [GodotBenchmark]
    public void ConcurrentLfuTest() {
        for (int Counter = 0; Counter < Count; Counter++) {
            Get(ConcurrentLfu.GetOrAdd(TargetString, static (string TargetString) => (StringName)TargetString));
        }
    }
    [GodotBenchmark]
    public void ExpiringConcurrentLruTest() {
        for (int Counter = 0; Counter < Count; Counter++) {
            Get(ExpiringConcurrentLru.GetOrAdd(TargetString, static (string TargetString) => (StringName)TargetString));
        }
    }
}
```
</details>

**Benchmark results:**
| Method                         | Mean      | Error    | StdDev   | Median    | Gen0       | Gen1      | Allocated  |
|------------------------------- |----------:|---------:|---------:|----------:|-----------:|----------:|-----------:|
| NoneTest                       | 155.71 ms | 3.110 ms | 4.460 ms | 156.71 ms |  3000.0000 | 2750.0000 | 19201012 B |
| FusionCacheTest                |  35.05 ms | 0.690 ms | 0.612 ms |  34.77 ms | 12200.0000 |         - | 38400561 B |
| MemoryCacheTest                |  19.39 ms | 0.363 ms | 0.304 ms |  19.29 ms |  4062.5000 |         - | 12800442 B |
| MemoryCacheStaticTest          |  17.84 ms | 0.224 ms | 0.175 ms |  17.76 ms |          - |         - |      442 B |
| ConcurrentDictionaryTest       |  17.01 ms | 0.282 ms | 0.264 ms |  17.02 ms |  4062.5000 |         - | 12800442 B |
| ConcurrentDictionaryStaticTest |  16.80 ms | 0.150 ms | 0.133 ms |  16.83 ms |          - |         - |      442 B |
| DictionaryWithLockTest         |  17.03 ms | 0.323 ms | 0.345 ms |  17.06 ms |          - |         - |      442 B |
| ConcurrentLruTest              |  16.56 ms | 0.050 ms | 0.039 ms |  16.55 ms |          - |         - |      442 B |
| ConcurrentLfuTest              |  18.23 ms | 0.502 ms | 1.466 ms |  17.63 ms |          - |         - |    54102 B |
| ExpiringConcurrentLruTest      |  17.04 ms | 0.287 ms | 0.268 ms |  17.07 ms |          - |         - |      442 B |

**Conclusion**
I decided to pick a capacity of 1,000 to avoid storing a large number of StringNames and an expire-after-access of 30 seconds to avoid storing old and unused StringNames. These can be changed but I think these work nicely.
Like I said before, this caching can be easily avoided if desired by using the constructor: `new StringName("...")`.
The result of this caching is a massive improvement in not only performance (9x faster) but also memory usage (19 MB -> 442 B), all while keeping the convenient implicit casting.

Joy-less commented:
This can also be emulated before the pull request is accepted using the following helper functions:
<details>
<summary>Open me</summary>

```cs
global using static RecycleExtensions;

using System.Diagnostics.CodeAnalysis;
using BitFaster.Caching;
using BitFaster.Caching.Lru;

public static class RecycleExtensions {
    private static readonly ICache<string, StringName> StringNameCache = new ConcurrentLruBuilder<string, StringName>()
        .WithCapacity(1_000)
        .WithExpireAfterAccess(TimeSpan.FromSeconds(30))
        .Build();
    private static readonly ICache<string, NodePath> NodePathCache = new ConcurrentLruBuilder<string, NodePath>()
        .WithCapacity(1_000)
        .WithExpireAfterAccess(TimeSpan.FromSeconds(30))
        .Build();

    /// <summary>
    /// Converts a <see cref="string"/> to a <see cref="Godot.StringName"/>.<br/>
    /// The resulting <see cref="Godot.StringName"/> is temporarily cached for future casts.
    /// </summary>
    [return: NotNullIfNotNull(nameof(String))]
    public static StringName? StringName(string? String) {
        if (String is null) {
            return null;
        }
        return StringNameCache.GetOrAdd(String, static (string From) => new StringName(From));
    }
    /// <summary>
    /// Converts a <see cref="string"/> to a <see cref="Godot.NodePath"/>.<br/>
    /// The resulting <see cref="Godot.NodePath"/> is temporarily cached for future casts.
    /// </summary>
    [return: NotNullIfNotNull(nameof(String))]
    public static NodePath? NodePath(string? String) {
        if (String is null) {
            return null;
        }
        return NodePathCache.GetOrAdd(String, static (string From) => new NodePath(From));
    }
}
```
</details>

Use like so:
```cs
StringName Apple = StringName("apple");
StringName Banana = NodePath("banana");
```

beicause commented on modules/mono/glue/GodotSharp/GodotSharp/Core/NodePath.cs:
```diff
@@ -48,6 +50,11 @@ public sealed class NodePath : IDisposable, IEquatable<NodePath?>
 
         private WeakReference<IDisposable>? _weakReferenceToSelf;
 
+        private static readonly ICache<string, NodePath> _nodePathCache = new ConcurrentLruBuilder<string, NodePath>()
+            .WithCapacity(1_000)
+            .WithExpireAfterAccess(TimeSpan.FromSeconds(30))
+            .Build();
+
```
This caching strategy may not be suitable for everyone, perhaps it can be public for users to manage on their own? And using ConcurrentDictionary may be sufficient instead of introducing additional dependencies.

Joy-less replied:
It could definitely be public and configurable but I thought it might be overkill. StringNames don't add much in C# so the ideal choice would be to axe them altogether in the long term.

As for using ConcurrentDictionary, that doesn't have eviction so it would cause memory leaks if StringNames were created dynamically (e.g. `StringName number = ("num" + i);`).

beicause replied:
The  `StringName`  inside Godot is reference-counted. So I think of a new idea. Perhaps we can also do reference counting for  StringName  in C# ?

beicause replied:
Well, it seems that can't be achieved.

beicause replied:
Hey, what about [WeakReference](https://learn.microsoft.com/en-us/dotnet/api/system.weakreference?view=net-9.0&redirectedfrom=MSDN), do you think it's suitable to use it as cache for this use case?

Joy-less replied:
`WeakReference` can't work as-is because it only applies to a single reference. You would have to use a `ConditionalWeakTable` or a collection. `ConditionalWeakTable`s don't really work with `string` keys. There are some interesting approaches with collections though:
- Create a fixed-size cache of `StringName` `WeakReference`s (e.g. `WeakReference<StringName>[1000]`) and do a linear search to find a matching `StringName`. This would be extremely slow.
- Use a `ConcurrentDictionary<string, WeakReference<StringName>` and clear the reference in the finalizer of `StringName`. I think this could work but there might be some issues with garbage collection.

Joy-less replied:
Something like this maybe:

```cs
private static readonly ConcurrentDictionary<string, WeakReference<StringName>> StringNames;

public static implicit operator StringName(string From) {
    return StringNames.GetOrAdd(From, From => new WeakReference<StringName>(new StringName(From)));
}

~StringName() {
    StringNames.TryRemove(ToString(), out _);
}
```

I am sceptical that this would provide good performance though. Creating `WeakReference`s is probably expensive compared to the approach with `BitFaster.Caching`, which by the way is an extremely micro-optimised library using vectorisation and unsafe code.

Joy-less replied:
I did some testing with the suggested approach and it yielded inconsistent results. `BitFaster.Caching` always gave better performance.

Joy-less commented:
I removed the `WithCapacity(1_000)` because I thought it's too much of a magic number in comparison to `WithExpireAfterAccess(TimeSpan.FromSeconds(30))`, and I made the package private so it's not visible to Godot projects.

Delsin-Yu commented:
Is there a solid reason we should invalidate the caches by time?

Joy-less commented:
> Is there a solid reason why we should invalidate the caches by time?
@Delsin-Yu

The caches must be invalidated at some point to prevent Ruby's [symbols denial-of-service](https://stackoverflow.com/questions/24066556/getting-warning-denial-of-service) problem (at least before mortal symbols were introduced).
As for the method of invalidation, time-based is the best solution I could find. Note that it invalidates 30s after the last access not after caching. The PR mainly prevents situations where you're creating a new StringName or NodePath every frame; if it's not used for 30s then it won't impact performance to expunge it from the cache. Ideally, it would take into account the amount of available memory, but there doesn't seem to be a good way to do this. I think the approach I suggested works for 99.99% of cases and is easy to workaround if someone wants to use their own caching system.

Delsin-Yu commented:
> if it's not used for 30s then it won't impact performance to expunge it from the cache.

It almost sounds like your solution aims to address the use case where the engine user suddenly creates an unreasonable number of StringNames (or NodePaths). Is this common in a typical project?

> The PR mainly prevents situations where you're creating a new StringName or NodePath every frame;

Generally, the tiered GC (as opposed to the Unity's legacy non-moving non-tiered GC) can take care of a small number (<500? benchmark needed) of these creations (or not to create stuttering), as tier 0 object GC is generally very fast.

Joy-less commented:
> It almost sounds like your solution aims to address the use case where the engine user suddenly creates an unreasonable number of StringNames (or NodePaths). Is this common in a typical project?

Very common, any project which uses them for getting nodes, checking input actions, setting shader parameters, or using GDScript variables will benefit from this.

> Generally, the tiered GC (as opposed to the Unity's legacy non-moving non-tiered GC) can take care of a small number (<500? benchmark needed) of these creations (or not to create stuttering), as tier 0 object GC is generally very fast.

Yes, C# is performant, but that doesn't mean it makes sense to make hidden allocations every frame for no reason.

beicause commented:
I'm curious how the performance would be if you used a lower number instead of 200,000 in previous benchmarks.

Delsin-Yu commented:
> that doesn't mean it makes sense to make hidden allocations every frame for no reason.

I believe we need to first prove the performance issue exists before addressing them.

> to make hidden allocations every frame for no reason.

The documentation [did tell you](https://docs.godotengine.org/en/stable/tutorials/scripting/c_sharp/c_sharp_basics.html#current-gotchas-and-known-issues) to use the cached StringNames.

Joy-less commented:
The benchmark project as requested by @Delsin-Yu:
[stringnamestest.zip](https://github.com/user-attachments/files/19462900/stringnamestest.zip)

Delsin-Yu commented:
The test numbers are here:

In general, any form of caching starts to make a difference when the numbers have reached `50_000` (on my end), so it helps more on eliminating GC spikes in common use cases.

<details>

```
BenchmarkDotNet v0.13.12, Windows 10 (10.0.19044.5608/21H2/November2021Update)
AMD Ryzen 9 5900X, 1 CPU, 24 logical and 12 physical cores
.NET SDK 9.0.200
  [Host] : .NET 9.0.2 (9.0.225.6610), X64 RyuJIT AVX2

Toolchain=InGodotProcessNoEmitToolchain  
```

| Method                         | Count  | Mean      | Error    | StdDev   | Ratio | RatioSD | Gen0      | Gen1     | Allocated  | Alloc Ratio |
|------------------------------- |------- |----------:|---------:|---------:|------:|--------:|----------:|---------:|-----------:|------------:|
| **NoneTest**                       | **100**    |  **16.58 ms** | **0.027 ms** | **0.023 ms** |  **1.00** |    **0.00** |         **-** |        **-** |    **10000 B** |        **1.00** |
| FusionCacheTest                | 100    |  16.60 ms | 0.011 ms | 0.010 ms |  1.00 |    0.00 |         - |        - |    19631 B |        1.96 |
| MemoryCacheTest                | 100    |  16.59 ms | 0.052 ms | 0.041 ms |  1.00 |    0.00 |         - |        - |     6831 B |        0.68 |
| MemoryCacheStaticTest          | 100    |  16.60 ms | 0.030 ms | 0.027 ms |  1.00 |    0.00 |         - |        - |      431 B |        0.04 |
| ConcurrentDictionaryStaticTest | 100    |  16.60 ms | 0.023 ms | 0.020 ms |  1.00 |    0.00 |         - |        - |      431 B |        0.04 |
| DictionaryWithLockTest         | 100    |  16.60 ms | 0.014 ms | 0.013 ms |  1.00 |    0.00 |         - |        - |      431 B |        0.04 |
| ConcurrentLruTest              | 100    |  16.59 ms | 0.032 ms | 0.026 ms |  1.00 |    0.00 |         - |        - |      431 B |        0.04 |
| ConcurrentLfuTest              | 100    |  16.59 ms | 0.022 ms | 0.018 ms |  1.00 |    0.00 |         - |        - |      493 B |        0.05 |
| ExpiringConcurrentLruTest      | 100    |  16.59 ms | 0.015 ms | 0.013 ms |  1.00 |    0.00 |         - |        - |      431 B |        0.04 |
|                                |        |           |          |          |       |         |           |          |            |             |
| **NoneTest**                       | **5000**   |  **16.46 ms** | **0.049 ms** | **0.043 ms** |  **1.00** |    **0.00** |         **-** |        **-** |   **480431 B** |       **1.000** |
| FusionCacheTest                | 5000   |  16.53 ms | 0.022 ms | 0.019 ms |  1.00 |    0.00 |   31.2500 |        - |   960431 B |       1.999 |
| MemoryCacheTest                | 5000   |  16.53 ms | 0.035 ms | 0.029 ms |  1.00 |    0.00 |         - |        - |   320431 B |       0.667 |
| MemoryCacheStaticTest          | 5000   |  16.57 ms | 0.016 ms | 0.015 ms |  1.01 |    0.00 |         - |        - |      431 B |       0.001 |
| ConcurrentDictionaryStaticTest | 5000   |  16.56 ms | 0.020 ms | 0.018 ms |  1.01 |    0.00 |         - |        - |      431 B |       0.001 |
| DictionaryWithLockTest         | 5000   |  16.57 ms | 0.020 ms | 0.017 ms |  1.01 |    0.00 |         - |        - |      440 B |       0.001 |
| ConcurrentLruTest              | 5000   |  16.56 ms | 0.040 ms | 0.034 ms |  1.01 |    0.00 |         - |        - |      431 B |       0.001 |
| ConcurrentLfuTest              | 5000   |  16.55 ms | 0.033 ms | 0.029 ms |  1.01 |    0.00 |         - |        - |     3353 B |       0.007 |
| ExpiringConcurrentLruTest      | 5000   |  16.55 ms | 0.044 ms | 0.037 ms |  1.01 |    0.00 |         - |        - |      431 B |       0.001 |
|                                |        |           |          |          |       |         |           |          |            |             |
| **NoneTest**                       | **10000**  |  **16.76 ms** | **0.035 ms** | **0.033 ms** |  **1.00** |    **0.00** |   **31.2500** |        **-** |   **960431 B** |       **1.000** |
| FusionCacheTest                | 10000  |  16.52 ms | 0.011 ms | 0.010 ms |  0.99 |    0.00 |   93.7500 |        - |  1920431 B |       2.000 |
| MemoryCacheTest                | 10000  |  16.51 ms | 0.029 ms | 0.024 ms |  0.99 |    0.00 |   31.2500 |        - |   640431 B |       0.667 |
| MemoryCacheStaticTest          | 10000  |  16.56 ms | 0.011 ms | 0.010 ms |  0.99 |    0.00 |         - |        - |      431 B |       0.000 |
| ConcurrentDictionaryStaticTest | 10000  |  16.55 ms | 0.031 ms | 0.026 ms |  0.99 |    0.00 |         - |        - |      431 B |       0.000 |
| DictionaryWithLockTest         | 10000  |  16.56 ms | 0.007 ms | 0.006 ms |  0.99 |    0.00 |         - |        - |      431 B |       0.000 |
| ConcurrentLruTest              | 10000  |  16.54 ms | 0.020 ms | 0.018 ms |  0.99 |    0.00 |         - |        - |      431 B |       0.000 |
| ConcurrentLfuTest              | 10000  |  16.52 ms | 0.068 ms | 0.053 ms |  0.99 |    0.00 |         - |        - |     6313 B |       0.007 |
| ExpiringConcurrentLruTest      | 10000  |  16.53 ms | 0.023 ms | 0.021 ms |  0.99 |    0.00 |         - |        - |      431 B |       0.000 |
|                                |        |           |          |          |       |         |           |          |            |             |
| **NoneTest**                       | **50000**  |  **47.06 ms** | **0.927 ms** | **1.764 ms** |  **1.00** |    **0.00** |  **272.7273** | **181.8182** |  **4800612 B** |       **1.000** |
| FusionCacheTest                | 50000  |  47.79 ms | 0.543 ms | 0.508 ms |  1.02 |    0.04 |  545.4545 |        - |  9600612 B |       2.000 |
| MemoryCacheTest                | 50000  |  20.41 ms | 0.320 ms | 0.299 ms |  0.44 |    0.02 |  187.5000 |        - |  3200440 B |       0.667 |
| MemoryCacheStaticTest          | 50000  |  19.49 ms | 0.269 ms | 0.252 ms |  0.42 |    0.02 |         - |        - |      431 B |       0.000 |
| ConcurrentDictionaryStaticTest | 50000  |  18.21 ms | 0.235 ms | 0.220 ms |  0.39 |    0.02 |         - |        - |      431 B |       0.000 |
| DictionaryWithLockTest         | 50000  |  18.26 ms | 0.270 ms | 0.252 ms |  0.39 |    0.02 |         - |        - |      431 B |       0.000 |
| ConcurrentLruTest              | 50000  |  19.40 ms | 0.229 ms | 0.214 ms |  0.42 |    0.02 |         - |        - |      431 B |       0.000 |
| ConcurrentLfuTest              | 50000  |  20.46 ms | 0.078 ms | 0.073 ms |  0.44 |    0.02 |         - |        - |    29586 B |       0.006 |
| ExpiringConcurrentLruTest      | 50000  |  19.33 ms | 0.225 ms | 0.210 ms |  0.41 |    0.01 |         - |        - |      431 B |       0.000 |
|                                |        |           |          |          |       |         |           |          |            |             |
| **NoneTest**                       | **100000** |  **91.81 ms** | **1.829 ms** | **3.480 ms** |  **1.00** |    **0.00** |  **500.0000** | **333.3333** |  **9600843 B** |       **1.000** |
| FusionCacheTest                | 100000 |  96.25 ms | 1.088 ms | 1.017 ms |  1.07 |    0.03 | 1000.0000 |        - | 19200843 B |       2.000 |
| MemoryCacheTest                | 100000 |  40.41 ms | 0.485 ms | 0.454 ms |  0.45 |    0.02 |  307.6923 |        - |  6400570 B |       0.667 |
| MemoryCacheStaticTest          | 100000 |  38.75 ms | 0.358 ms | 0.335 ms |  0.43 |    0.01 |         - |        - |      570 B |       0.000 |
| ConcurrentDictionaryStaticTest | 100000 |  36.57 ms | 0.308 ms | 0.288 ms |  0.41 |    0.01 |         - |        - |      553 B |       0.000 |
| DictionaryWithLockTest         | 100000 |  36.30 ms | 0.484 ms | 0.453 ms |  0.40 |    0.01 |         - |        - |      553 B |       0.000 |
| ConcurrentLruTest              | 100000 |  38.71 ms | 0.432 ms | 0.404 ms |  0.43 |    0.01 |         - |        - |      553 B |       0.000 |
| ConcurrentLfuTest              | 100000 |  40.96 ms | 0.314 ms | 0.279 ms |  0.45 |    0.02 |         - |        - |    59518 B |       0.006 |
| ExpiringConcurrentLruTest      | 100000 |  38.66 ms | 0.529 ms | 0.495 ms |  0.43 |    0.01 |         - |        - |      553 B |       0.000 |
|                                |        |           |          |          |       |         |           |          |            |             |
| **NoneTest**                       | **200000** | **182.50 ms** | **3.631 ms** | **6.454 ms** |  **1.00** |    **0.00** | **1000.0000** | **666.6667** | **19201349 B** |       **1.000** |
| FusionCacheTest                | 200000 | 190.42 ms | 2.502 ms | 2.341 ms |  1.05 |    0.04 | 2000.0000 |        - | 38401349 B |       2.000 |
| MemoryCacheTest                | 200000 |  80.64 ms | 0.756 ms | 0.670 ms |  0.45 |    0.01 |  714.2857 |        - | 12800770 B |       0.667 |
| MemoryCacheStaticTest          | 200000 |  77.39 ms | 1.113 ms | 1.041 ms |  0.43 |    0.02 |         - |        - |      770 B |       0.000 |
| ConcurrentDictionaryStaticTest | 200000 |  73.00 ms | 0.997 ms | 0.933 ms |  0.40 |    0.01 |         - |        - |      770 B |       0.000 |
| DictionaryWithLockTest         | 200000 |  73.37 ms | 1.442 ms | 1.349 ms |  0.41 |    0.02 |         - |        - |      770 B |       0.000 |
| ConcurrentLruTest              | 200000 |  78.51 ms | 1.490 ms | 1.464 ms |  0.43 |    0.01 |         - |        - |      770 B |       0.000 |
| ConcurrentLfuTest              | 200000 |  82.03 ms | 0.554 ms | 0.491 ms |  0.46 |    0.02 |         - |        - |   118216 B |       0.006 |
| ExpiringConcurrentLruTest      | 200000 |  76.77 ms | 0.999 ms | 0.935 ms |  0.43 |    0.01 |         - |        - |      770 B |       0.000 |

</details>

You may grab the test project an try it yourself: 

[stringnamestest.zip](https://github.com/user-attachments/files/19464089/stringnamestest.zip)

This project also includes the ability to test GC spike interactively:

https://github.com/user-attachments/assets/5beae9b7-4305-446c-978d-6d6f7d88cde7

beicause commented:
I believe caching can bring enough performance improvement, but I wonder if it's possible to implement a more fundamental solution to reduce the overhead of implicit conversion.
Compared with the implicit conversion of `StringName` in GDScript, how much slower it is in C#?

Joy-less commented:
> I believe caching can bring enough performance improvement, but I wonder if it's possible to implement a more fundamental solution to reduce the overhead of implicit conversion.
> Compared with the implicit conversion of `StringName` in GDScript, how much slower it is in C#?

In GDScript it's not as big of a problem for three reasons:
1. StringNames/NodePaths are built into the language, so you have `&StringName` and `^NodePath` syntax and also the parser can detect implicit casts.
2. GDScript uses reference counting which results in slower performance but not garbage collection spikes. (I think)
3. GDScript is closer to C++ so it only has to maintain one StringName (rather than marshalling a C# copy)

beicause commented:
This is a patch using WeakReference and ConcurrentDictionary as cache:
```diff
diff --git a/modules/mono/glue/GodotSharp/GodotSharp/Core/NodePath.cs b/modules/mono/glue/GodotSharp/GodotSharp/Core/NodePath.cs
index 0af640533d..0dbcce43de 100644
--- a/modules/mono/glue/GodotSharp/GodotSharp/Core/NodePath.cs
+++ b/modules/mono/glue/GodotSharp/GodotSharp/Core/NodePath.cs
@@ -1,4 +1,5 @@
 using System;
+using System.Collections.Concurrent;
 using System.Diagnostics.CodeAnalysis;
 using Godot.NativeInterop;
 
@@ -48,8 +49,16 @@ namespace Godot
 
         private WeakReference<IDisposable>? _weakReferenceToSelf;
 
+        private string? _cache_key;
+
+        /// <summary>
+        ///
+        /// </summary>
+        public static readonly ConcurrentDictionary<string, WeakReference<NodePath>> NodePathCache = new();
+
         ~NodePath()
         {
+            if (_cache_key != null) NodePathCache.TryRemove(_cache_key, out _);
             Dispose(false);
         }
 
@@ -129,16 +138,32 @@ namespace Godot
         }
 
         /// <summary>
-        /// Converts a string to a <see cref="NodePath"/>.
+        /// Converts a <see cref="string"/> to a <see cref="NodePath"/>.<br/>
+        /// The resulting <see cref="NodePath"/> is temporarily cached for future casts.
         /// </summary>
         /// <param name="from">The string to convert.</param>
-        public static implicit operator NodePath(string from) => new NodePath(from);
+        public static implicit operator NodePath(string from)
+        {
+            if (NodePathCache.TryGetValue(from, out WeakReference<NodePath>? weakref) && weakref != null)
+            {
+                if (weakref.TryGetTarget(out NodePath? val) && val != null)
+                {
+                    return val;
+                }
+            }
+            var ret = new NodePath(from)
+            {
+                _cache_key = from
+            };
+            NodePathCache[from] = new(ret);
+            return ret;
+        }
 
         /// <summary>
-        /// Converts this <see cref="NodePath"/> to a string.
+        /// Converts a <see cref="NodePath"/> to a <see cref="string"/>.
         /// </summary>
         /// <param name="from">The <see cref="NodePath"/> to convert.</param>
-        [return: NotNullIfNotNull("from")]
+        [return: NotNullIfNotNull(nameof(from))]
         public static implicit operator string?(NodePath? from) => from?.ToString();
 
         /// <summary>
diff --git a/modules/mono/glue/GodotSharp/GodotSharp/Core/StringName.cs b/modules/mono/glue/GodotSharp/GodotSharp/Core/StringName.cs
index 21d9ada127..846094ee21 100644
--- a/modules/mono/glue/GodotSharp/GodotSharp/Core/StringName.cs
+++ b/modules/mono/glue/GodotSharp/GodotSharp/Core/StringName.cs
@@ -1,4 +1,5 @@
 using System;
+using System.Collections.Concurrent;
 using System.Diagnostics.CodeAnalysis;
 using Godot.NativeInterop;
 
@@ -19,8 +20,16 @@ namespace Godot
 
         private WeakReference<IDisposable>? _weakReferenceToSelf;
 
+        private string? _cache_key;
+
+        /// <summary>
+        ///
+        /// </summary>
+        public static readonly ConcurrentDictionary<string, WeakReference<StringName>> StringNameCache = new();
+
         ~StringName()
         {
+            if (_cache_key != null) StringNameCache.TryRemove(_cache_key, out _);
             Dispose(false);
         }
 
@@ -75,16 +84,32 @@ namespace Godot
         }
 
         /// <summary>
-        /// Converts a string to a <see cref="StringName"/>.
+        /// Converts a <see cref="string"/> to a <see cref="StringName"/>.<br/>
+        /// The resulting <see cref="StringName"/> is temporarily cached for future casts.
         /// </summary>
         /// <param name="from">The string to convert.</param>
-        public static implicit operator StringName(string from) => new StringName(from);
+        public static implicit operator StringName(string from)
+        {
+            if (StringNameCache.TryGetValue(from, out WeakReference<StringName>? weakref) && weakref != null)
+            {
+                if (weakref.TryGetTarget(out StringName? val) && val != null)
+                {
+                    return val;
+                }
+            }
+            var ret = new StringName(from)
+            {
+                _cache_key = from
+            };
+            StringNameCache[from] = new(ret);
+            return ret;
+        }
 
         /// <summary>
-        /// Converts a <see cref="StringName"/> to a string.
+        /// Converts a <see cref="StringName"/> to a <see cref="string"/>.
         /// </summary>
         /// <param name="from">The <see cref="StringName"/> to convert.</param>
-        [return: NotNullIfNotNull("from")]
+        [return: NotNullIfNotNull(nameof(from))]
         public static implicit operator string?(StringName? from) => from?.ToString();
 
         /// <summary>

```

The following two tests are in the release mode, where the improvement by cache is more significant. 
```
// * Summary *

BenchmarkDotNet v0.13.12, EndeavourOS
AMD Ryzen 7 4800U with Radeon Graphics, 1 CPU, 16 logical and 8 physical cores
.NET SDK 9.0.202
  [Host] : .NET 9.0.3 (9.0.325.11113), X64 RyuJIT AVX2

Toolchain=InGodotProcessNoEmitToolchain  

| Method                         | Count | Mean     | Error     | StdDev    | Ratio | RatioSD | Gen0     | Gen1     | Allocated | Alloc Ratio |
|------------------------------- |------ |---------:|----------:|----------:|------:|--------:|---------:|---------:|----------:|------------:|
| NoneTest                       | 10000 | 9.644 ms | 0.2711 ms | 0.7992 ms |  1.00 |    0.00 | 140.6250 | 125.0000 |  960373 B |       1.000 |
| ConcurrentDictionaryStaticTest | 10000 | 1.366 ms | 0.0284 ms | 0.0838 ms |  0.14 |    0.02 |        - |        - |     342 B |       0.000 |
```
```
// * Summary *

BenchmarkDotNet v0.13.12, EndeavourOS
AMD Ryzen 7 4800U with Radeon Graphics, 1 CPU, 16 logical and 8 physical cores
.NET SDK 9.0.202
  [Host] : .NET 9.0.3 (9.0.325.11113), X64 RyuJIT AVX2

Toolchain=InGodotProcessNoEmitToolchain  

| Method                         | Count | Mean     | Error     | StdDev    | Ratio | RatioSD | Allocated | Alloc Ratio |
|------------------------------- |------ |---------:|----------:|----------:|------:|--------:|----------:|------------:|
| ReuseTest                      | 10000 | 1.247 ms | 0.0655 ms | 0.1772 ms |  1.00 |    0.00 |     340 B |        1.00 |
| WeakRefTest                    | 10000 | 1.245 ms | 0.0249 ms | 0.0650 ms |  1.01 |    0.13 |     342 B |        1.01 |
| ConcurrentDictionaryStaticTest | 10000 | 1.335 ms | 0.0369 ms | 0.1071 ms |  1.09 |    0.18 |     342 B |        1.01 |
| ExpiringConcurrentLruTest      | 10000 | 1.467 ms | 0.0292 ms | 0.0706 ms |  1.20 |    0.16 |     342 B |        1.01 |
```
In my opinion, compared with time-based caching, WeakReference caching is good as well.

Joy-less commented:
@beicause Nice work! One possible problem with weak reference caching is that, if the garbage collector runs often, then string names / node paths could be collected and reallocated more often than they should. I think it could still be more ideal though since it will take into account the available memory of the system.

Joy-less commented:
@beicause I replaced the `BitFaster.Caching` dependency with weak reference caching like you suggested. I made some changes to your implementation:
1. Your use of `TryGetValue` and `[]=` is not thread-safe. I changed it to `GetOrAdd`.
2. I noticed you added allocation-free conversions to `string` which is good. I made it so if it has to take the allocation path when converting to a string, it sets the cached string so it doesn't have to do it again.

beicause commented:
Note that the `NodePath.ToString`  should not return the original string. The result may be different from the original one after NodePath parses the string

beicause commented:
After I merged #104689, I found the improvement by reusing StringName is not very significant anymore.

https://github.com/user-attachments/assets/b24778ff-a312-4602-998e-d21adf9c9148

beicause commented:
The following test proves that the bottleneck is creating the NativeValue, because creating an empty StringName will not call `godotsharp_string_name_new_from_string`. If possible, it would be better to directly cache the `godotsharp_string_name`. 
```cs
    [GodotBenchmark]
    public void NewEmptyStringNameTest()
    {
        for (var counter = 0; counter < Count; counter++) new StringName("");
    }

    [GodotBenchmark]
    public void NewStringNameTest()
    {
        for (var counter = 0; counter < Count; counter++) new StringName(_targetString);
    }
```

https://github.com/user-attachments/assets/32e9519d-6d8f-41ba-8ba6-7b32f81c104d

beicause commented on modules/mono/glue/GodotSharp/GodotSharp/Core/NodePath.cs:
```diff
@@ -150,10 +187,16 @@ public override string ToString()
             if (IsEmpty)
                 return string.Empty;
 
+            if (_stringRepresentation is not null)
+                return _stringRepresentation;
+
```
This is not correct. For example, `new NodePath("//").ToString()` should return "/", but this will return "//".

Joy-less replied:
Fixed this for NodePaths @beicause
At this point though, whether ToString uses a cached version is pretty unpredictable, would it make sense to always set `_stringRepresentation` in the constructor by calling the native function?

Joy-less commented:
I remade this pull request in a way that makes `StringName` and `NodePath` classes immutable, predictable, and thread-safe.

Here's how it works:
- Constructing a C# StringName or NodePath:
  - The constructor stores the input string in a private field (`_inputString`). This corresponds to the key in the cache.
  - The constructor stores the output string in a private field (`_outputString`). This string is calculated using the native value, or taken directly from the input string for StringNames since they can't be simplified.
- Converting a C# StringName or NodePath to a C# string:
  - The output string stored in `_outputString` is returned.
- Implicitly converting to a C# StringName or NodePath from a C# string:
  - There is an attempt to fetch an existing StringName or NodePath from the cache, using the input string as a key.
  - If not found, a new StringName or NodePath is created and returned. There is also an attempt to add it to the cache (which is ignored if this has already happened by another thread due to a race).
- Finalizing/disposing a C# StringName or NodePath:
  - The cache entry is removed if there is a cache entry matching the input string to the object being disposed.
- Comparing the equality of a C# StringName or NodePath and a C# string:
  - A new method (`Equals(string?)`) has been added. For StringNames, it compares the given string and `_outputString`. For NodePaths, it converts the given string to a native NodePath, compares the native handles, and disposes the new native NodePath (this should still avoid C# garbage collection).

The result:
- The fields of C# StringNames and NodePaths are readonly, making them immutable and therefore thread-safe.
- Converting StringNames and NodePaths to C# strings (e.g. through `.ToString()`) is free.
- Converting strings to StringNames and NodePaths via implicit conversion uses caching and therefore cheap.

~~One question:~~
~~- In the C# NodePath constructor that accepts a string, is it necessary to convert to/from a native handle to get the stringified version? In other words, is `new NodePath(from).ToString() == from` ever `false`?~~
Answer: yes it is, because `//` simplifies to `/`.

@beicause Could I get a review from you? Would be appreciated.

Delsin-Yu commented on modules/mono/glue/GodotSharp/GodotSharp/Core/StringName.cs:
```diff
@@ -59,6 +79,10 @@ internal static StringName CreateTakingOwnershipOfDisposableValue(godot_string_n
         /// </summary>
         public StringName()
         {
+            // Store input string (used to create NodePath)
```
```suggestion
            // Store input string (used to create this StringName)
```

Joy-less committed:
Cache StringNames & NodePaths (squash 2)

Joy-less commented:
Closing this pull request in favour of https://github.com/godotengine/godot-proposals/discussions/15007.
